Readme.md
# retext-lancaster-stemmer [![Build Status](https://img.shields.io/travis/wooorm/retext-lancaster-stemmer.svg?style=flat)](https://travis-ci.org/wooorm/retext-lancaster-stemmer) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-lancaster-stemmer.svg?style=flat)](https://coveralls.io/r/wooorm/retext-lancaster-stemmer?branch=master)

**[retext](https://github.com/wooorm/retext)** implementation of the [Lancaster (Paice/Husk) stemming algorithm](http://www.comp.lancs.ac.uk/computing/research/stemming/index.htm).

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-lancaster-stemmer
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-lancaster-stemmer
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-lancaster-stemmer
```

[Duo](http://duojs.org/#getting-started):

```javascript
var lancasterStemmer = require('wooorm/retext-lancaster-stemmer');
```

## Usage

```javascript
var Retext = require('retext');
var inspect = require('retext-inspect');
var visit = require('retext-visit');
var lancasterStemmer = require('retext-lancaster-stemmer');

var retext = new Retext()
    .use(inspect)
    .use(visit)
    .use(lancasterStemmer)

retext.parse('A simple English sentence.', function (err, tree) {
    tree.visit(tree.WORD_NODE, function (node) {
        console.log(node);
    });
    /**
     * WordNode[1] [data={"stem":"a"}]
     * └─ TextNode: 'A'
     * WordNode[1] [data={"stem":"simpl"}]
     * └─ TextNode: 'simple'
     * WordNode[1] [data={"stem":"engl"}]
     * └─ TextNode: 'English'
     * WordNode[1] [data={"stem":"sent"}]
     * └─ TextNode: 'sentence'
     */
});
```

## API

None, **retext-lancaster-stemmer** automatically detects the stem of each [`WordNode`](https://github.com/wooorm/textom#textomwordnode-nlcstwordnode) (using **[wooorm/stemmer](https://github.com/wooorm/lancaster-stemmer)**), and stores the stem in `node.data.stem`.

## Performance

On a MacBook Air, **retext** performs about 13% slower with **retext-lancaster-stemmer**.

```text
           retext w/o retext-lancaster-stemmer
  225 op/s » A paragraph (5 sentences, 100 words)
   25 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)

           retext w/ retext-lancaster-stemmer
  195 op/s » A paragraph (5 sentences, 100 words)
   20 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
```

## Related

- [retext-porter-stemmer](https://github.com/wooorm/retext-porter-stemmer) — Same workings, but using the Porter stemming algorithm.

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.1.7 / 2015-01-19
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update eslint
  * Refactor to use generic `changeinside` in new retext
  * Update retext to stable release

0.1.6 / 2014-12-13
==================

 * Add benchmark
 * Fix wording in `Readme.md`
 * Add link to personal website to copyright in `Readme.md`
 * Update eslint, retext

0.1.5 / 2014-11-21
==================

 * Refactor `Readme.md`
 * Add flat badges to `Readme.md`
 * Fix incorrect executive rights on `test.js`
 * Refactor npm script targets in `package.json`
 * Add retext-inspect to dev-dependecies
 * Update retext

0.1.4 / 2014-10-27
==================

 * Add `.eslintrc`
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update eslint, mocha, retext, retext-visit, retext-content

0.1.3 / 2014-10-21
==================

 * Refactor test
 * Refactor .jscs.json
 * Update retext

0.1.2 / 2014-10-14
==================

 * Update tests for changes to retext-visit
 * Update jscs, retext, retext-content, retext-visit
 * Update .gitignore, .npmignore, bower ignore
 * Move spec to test.js

0.1.1 / 2014-09-27
==================

 * Refactor Readme.md
 * Refactor API
 * Add npm deployment to travis
 * Remove `npm update` from travis
 * Remove browser test
 * Update .gitignore, .npmignore
 * Update Installation in docs
 * Add bower.json
 * Fix property order in component.json, package.json
 * Remove testling
 * Update spec
 * Remove `lancasterStemmer` from exports
 * Update dependencies


test.js
'use strict';

/*
 * Module dependencies.
 */

var Retext,
    stemmer,
    inspect,
    content,
    visit,
    assert;

Retext = require('retext');
stemmer = require('./');
inspect = require('retext-inspect');
content = require('retext-content');
visit = require('retext-visit');
assert = require('assert');

/*
 * Retext.
 */

var retext;

retext = new Retext()
    .use(inspect)
    .use(content)
    .use(visit)
    .use(stemmer);

/*
 * Constants.
 */

var otherWords,
    otherStems;

otherWords = ['An', 'easy', 'normal', 'paragraph'];
otherStems = ['an', 'easy', 'norm', 'paragraph'];

/*
 * Tests.
 */

describe('porterStemmer()', function () {
    it('should be a `function`', function () {
        assert(typeof stemmer === 'function');
    });

    retext.parse('A simple, english, sentence', function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should `stem` each `WordNode`', function () {
            tree.visit(tree.WORD_NODE, function (wordNode) {
                assert('stem' in wordNode.data);
            });
        });

        it('should set each stem to `null` when a WordNode (no longer?) ' +
            'has a value', function () {
                tree.visit(tree.WORD_NODE, function (wordNode) {
                    wordNode.removeContent();

                    assert(wordNode.data.stem === null);
                });
            }
        );

        it('should automatically re-stem `WordNode`s when their values ' +
            'change',
            function () {
                var index;

                index = -1;

                tree.visit(tree.WORD_NODE, function (wordNode) {
                    wordNode.replaceContent(otherWords[++index]);

                    assert(wordNode.data.stem === otherStems[index]);
                });
            }
        );
    });
});


spec/retext-lancaster-stemmer.spec.js
'use strict';

var stemmer,
    content,
    visit,
    Retext,
    assert,
    retext,
    otherWords,
    otherStems;

/**
 * Module dependencies.
 */

stemmer = require('..');
Retext = require('retext');
visit = require('retext-visit');
content = require('retext-content');
assert = require('assert');

/**
 * Retext.
 */

retext = new Retext()
    .use(visit)
    .use(content)
    .use(stemmer);

/**
 * Constants.
 */

otherWords = ['An', 'easy', 'normal', 'paragraph'];
otherStems = ['an', 'easy', 'norm', 'paragraph'];

/**
 * Tests.
 */

describe('porterStemmer()', function () {
    it('should be a `function`', function () {
        assert(typeof stemmer === 'function');
    });

    it('should export an `attach` method', function () {
        assert(typeof stemmer.attach === 'function');
    });

    retext.parse('A simple, english, sentence', function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should `stem` each `WordNode`', function () {
            tree.visitType(tree.WORD_NODE, function (wordNode) {
                assert('stem' in wordNode.data);
            });
        });

        it('should set each stem to `null` when a WordNode (no longer?) ' +
            'has a value', function () {
                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    wordNode.removeContent();

                    assert(wordNode.data.stem === null);
                });
            }
        );

        it('should automatically re-stem `WordNode`s when their values ' +
            'change',
            function () {
                var index;

                index = -1;

                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    wordNode.replaceContent(otherWords[++index]);

                    assert(wordNode.data.stem === otherStems[index]);
                });
            }
        );
    });
});


index.js
/**
 * Dependencies.
 */

var Retext = require('wooorm/retext@0.5.0');
var lancasterStemmer = require('wooorm/retext-lancaster-stemmer@0.1.7');
var dom = require('wooorm/retext-dom@0.3.2');
var visit = require('wooorm/retext-visit@0.2.5');

/**
 * Retext.
 */

var retext = new Retext()
    .use(visit)
    .use(dom)
    .use(lancasterStemmer);

/**
 * DOM elements.
 */

var $input = document.getElementsByTagName('textarea')[0];
var $output = document.getElementsByTagName('div')[0];

/**
 * Get a color representation from a string.
 */

var style = document.styleSheets[0];

var stems = {};

function hashCode(str) {
    var hash = 0;
    for (var i = 0; i < str.length; i++) {
       hash = str.charCodeAt(i) + ((hash << 5) - hash);
    }
    return hash;
}

function intToARGB(i){
    return ((i>>24)&0xFF).toString(16) +
           ((i>>16)&0xFF).toString(16) +
           ((i>>8)&0xFF).toString(16) +
           (i&0xFF).toString(16);
}

function getColorFromString(value) {
    value = intToARGB(hashCode(value)).slice(0, 6);

    while (value.length < 6) {
        value = '0' + value;
    }

    return '#' + value;
}

/**
 * Add a CSS rule.
 */

function addCSSRule(sheet, selector, rules) {
    if(sheet.insertRule) {
        sheet.insertRule(selector + '{' + rules + '}');
    } else {
        sheet.addRule(selector, rules);
    }
}

/**
 * Callback when new stems are calculated.
 */

function onstem(stem) {
    var color;

    if (stem in stems) {
        return;
    }

    color = getColorFromString(stem)
    stems[stem] = color;

    addCSSRule(style, '[title="' + stem + '"]', 'color:' + color);
}

/**
 * Events
 */

var tree;

function oninputchange() {
    if (tree) {
        tree.toDOMNode().parentNode.removeChild(tree.toDOMNode());
    }

    retext.parse($input.value, function (err, root) {
        if (err) throw err;

        tree = root;

        tree.visit(function (node) {
            if (!node.DOMTagName || !node.data.stem) {
                return;
            }

            stem = node.data.stem;

            onstem(stem);

            node.toDOMNode().setAttribute('title', stem);
        });

        $output.appendChild(tree.toDOMNode());
    });
}

$input.addEventListener('input', oninputchange);

oninputchange();


benchmark.js
'use strict';

var Retext,
    lancasterStemmer;

/*
 * Dependencies.
 */

Retext = require('retext');
lancasterStemmer = require('./');

/*
 * Dependencies.
 */

var retext,
    retextWithLancasterStemmer;

retext = new Retext();
retextWithLancasterStemmer = new Retext().use(lancasterStemmer);

/*
 * Test data: A (big?) article (w/ 100 paragraphs, 500
 * sentences, 10,000 words);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

suite('retext w/o retext-lancaster-stemmer', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retext.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retext.parse(section, done);
        }
    );
});

suite('retext w/ retext-lancaster-stemmer', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retextWithLancasterStemmer.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retextWithLancasterStemmer.parse(section, done);
        }
    );
});
